mm2src/lp_ordermatch/simple_market_maker.rs

```diff
@@ -0,0 +1,702 @@
+pub enum StopSimpleMakerBotError {
+    #[display(fmt = "The bot is already stopped")]
+    AlreadyStopped,
+    #[display(fmt = "The bot is already stopping")]
+    AlreadyStopping,
+    #[display(fmt = "Transport error: {}", _0)]
+    Transport(String),
+    #[display(fmt = "Internal error: {}", _0)]
+    InternalError(String),
+}
+
+#[derive(Debug, Deserialize, Display, Serialize, SerializeErrorType)]
+#[serde(tag = "error_type", content = "error_data")]
+pub enum StartSimpleMakerBotError {
+    #[display(fmt = "The bot is already started")]
+    AlreadyStarted,
+    #[display(fmt = "Invalid bot configuration")]
+    InvalidBotConfiguration,
+    #[display(fmt = "Transport error: {}", _0)]
+    Transport(String),
+    #[display(fmt = "Internal error: {}", _0)]
+    InternalError(String),
+}
+
+#[derive(Debug)]
+pub enum PriceServiceRequestError {
+    HttpProcessError(String),
+    ParsingAnswerError(String),
+}
+
+impl From<std::string::String> for PriceServiceRequestError {
+    fn from(error: String) -> Self { PriceServiceRequestError::HttpProcessError(error) }
+}
+
+impl From<std::str::Utf8Error> for PriceServiceRequestError {
+    fn from(error: Utf8Error) -> Self { PriceServiceRequestError::HttpProcessError(error.to_string()) }
+}
+
+impl HttpStatusCode for StartSimpleMakerBotError {
+    fn status_code(&self) -> StatusCode {
+        match self {
+            StartSimpleMakerBotError::AlreadyStarted | StartSimpleMakerBotError::InvalidBotConfiguration => {
+                StatusCode::BAD_REQUEST
+            },
+            StartSimpleMakerBotError::Transport(_) | StartSimpleMakerBotError::InternalError(_) => {
+                StatusCode::INTERNAL_SERVER_ERROR
+            },
+        }
+    }
+}
+
+impl HttpStatusCode for StopSimpleMakerBotError {
+    fn status_code(&self) -> StatusCode {
+        match self {
+            // maybe bad request is not adapted for the first errors.
+            StopSimpleMakerBotError::AlreadyStopped | StopSimpleMakerBotError::AlreadyStopping => {
+                StatusCode::BAD_REQUEST
+            },
+            StopSimpleMakerBotError::Transport(_) | StopSimpleMakerBotError::InternalError(_) => {
+                StatusCode::INTERNAL_SERVER_ERROR
+            },
+        }
+    }
+}
+
+struct TradingPair {
+    base: String,
+    rel: String,
+}
+
+impl TradingPair {
+    pub fn new(base: String, rel: String) -> TradingPair { TradingPair { base, rel } }
+
+    pub fn as_combination(&self) -> String { self.base.clone() + "/" + self.rel.clone().as_str() }
+}
+
+pub async fn tear_down_bot(ctx: MmArc) {
+    let simple_market_maker_bot_ctx = TradingBotContext::from_ctx(&ctx).unwrap();
+    {
+        let mut trading_bot_cfg = simple_market_maker_bot_ctx.trading_bot_cfg.lock().await;
+        trading_bot_cfg.clear();
+    }
+    cancel_pending_orders(&ctx).await;
+}
+
+async fn get_non_zero_balance(coin: MmCoinEnum) -> Result<MmNumber, MmError<OrderProcessingError>> {
```

Comment:
`OrderProcessingError` has many variants, which won't be ever created by `get_non_zero_balance`. Please consider adding a separate error type for this.